src/server.ts

```diff
@@ -81,66 +83,178 @@ export async function buildServer (ctx: ViteBuildContext) {
-        .replace('{{ APP }}', '<div id="__nuxt">{{ APP }}</div>')
-        .replace(
-          '</body>',
-          '<script type="module" src="/@vite/client"></script><script type="module" src="/client.js"></script></body>'
-        )
     : DEFAULT_APP_TEMPLATE
 
-  await writeFile(resolve(serverDist, 'index.ssr.html'), APP_TEMPLATE)
-  await writeFile(resolve(serverDist, 'index.spa.html'), APP_TEMPLATE)
-  await writeFile(resolve(serverDist, 'client.manifest.json'), JSON.stringify({
-    publicPath: '',
-    all: [],
+  const SPA_TEMPLATE = APP_TEMPLATE
+    .replace('{{ APP }}', '<div id="__nuxt">{{ APP }}</div>')
+    .replace(
+      '</body>',
+      '<script type="module" src="/@vite/client"></script><script type="module" src="/client.js"></script></body>'
+    )
+  const SSR_TEMPLATE = ctx.nuxt.options.dev ? SPA_TEMPLATE : APP_TEMPLATE
+
+  await writeFile(r('index.ssr.html'), SSR_TEMPLATE)
+  await writeFile(r('index.spa.html'), SPA_TEMPLATE)
+
+  if (ctx.nuxt.options.dev) {
+    await stubManifest(ctx)
+  } else {
+    await generateBuildManifest(ctx)
+  }
+
+  const onBuild = () => ctx.nuxt.callHook('build:resources', wpfs)
+  const build = async () => {
+    const start = Date.now()
+    await vite.build(serverConfig)
+    await onBuild()
+    consola.info(`Server built in ${Date.now() - start}ms`)
+  }
+
+  const debouncedBuild = debounce(build, 300)
+
+  await build()
+
+  if (ctx.nuxt.options.dev) {
+    const watcher = watch([
+      ctx.nuxt.options.buildDir,
+      ctx.nuxt.options.srcDir,
+      ctx.nuxt.options.rootDir
+    ], {
+      ignored: [
+        '**/dist/server/**'
+      ]
+    })
+
+    watcher.on('change', () => debouncedBuild())
+
+    ctx.nuxt.hook('close', async () => {
+      await watcher.close()
+    })
+  }
+}
+
+// convert vite's manifest to webpack style
+async function generateBuildManifest (ctx: ViteBuildContext) {
+  const serverDist = resolve(ctx.nuxt.options.buildDir, 'dist/server')
+  const clientDist = resolve(ctx.nuxt.options.buildDir, 'dist/client')
+  const r = (...args: string[]): string => resolve(serverDist, ...args)
+
+  const publicPath = ctx.nuxt.options.app.assetsPath // Default: /nuxt/
+  const viteClientManifest = await readJSON(join(clientDist, 'manifest.json'))
+
+  function getModuleIds ([, value]: [string, any]) {
+    if (!value) {
+      return []
+    }
+    return [value.file, ...value.css || []]
+      .filter(i => !i.endsWith('.js') || i.match(/-legacy\./)) // only use legacy build
+  }
+
+  const asyncEntries = uniq(
+    Object.entries(viteClientManifest)
+      .filter((i: any) => i[1].isDynamicEntry)
+      .flatMap(getModuleIds)
+  ).filter(i => i)
+  const initialEntries = uniq(
+    Object.entries(viteClientManifest)
+      .filter((i: any) => !i[1].isDynamicEntry)
+      .flatMap(getModuleIds)
+  ).filter(i => i)
+  const initialJs = initialEntries.filter(i => i.endsWith('.js'))
+  const initialAssets = initialEntries.filter(i => !i.endsWith('.js'))
```

Comment:
Refactor notes regarding `generateBuildManifest`: We have kinda duplicate logic in vue-bundle-renderer that can be merged like normalizing manifest (https://github.com/nuxt-contrib/vue-bundle-renderer/pull/15) and `isJS`, `isCSS`  utils